### ISSUE TITLE
Sources never copied when source and binary package names differ

### ISSUE THREAD

**Issue by benjaoming** [mgedmin/ppa-copy-packages#2]
THANKS FOR A GREAT SCRIPT! And thanks for packaging it and sharing it after all these years! It's saved me a lot of time and clicks :)

Can you confirm if the script is currently working for you?

The source dist and its binaries are all built and ready. `ppa-copy-packages` are running without errors. This is the output:

```
➜  debian git:(0.17.6b6) ✗ ppa-copy-packages --owner learningequality --name ka-lite-proposed -p ka-lite-source -s trusty -t focal --verbose                  
ka-lite-source 0.17.6~b10-0ubuntu1 missing from focal
  but binaries aren't published yet (state: Successfully built) - https://launchpad.net/~learningequality/+archive/ubuntu/ka-lite-proposed/+build/19373919
```

I tried the Launchpad interface and copying worked fine and subsequently rendered a correct error in `ppa-copy-packages`.

Is the deprecation perhaps the issue?

https://github.com/mgedmin/ppa-copy-packages/blob/d4de9fca8ccb1147dee4a31ad628977d4553fe28/ppa_copy_packages.py#L312-L323

**Comment by mgedmin** [comment]
Yes, the script is working for me, or at least it was back at the end of April when I used it to prepare both of my PPAs for focal.

Note this message:

>  but binaries aren't published yet

This means you have to wait for the cron script that updates package index files for your PPA before Launchpad will allow the copy.  I think the cron script runs every 20 minutes, but don't quote me on that.  That's why ppa-copy-packages has the -w (--wait) flag.

I don't understand what you meant when you said

> subsequently rendered a correct error in ppa-copy-packages.

?

Also, I'm curious that trying from the web UI allowed the copy -- could it be a coincidence, the package getting published after you tried ppa-copy-packages but before you tried the manual copy?  

Otherwise, if Launchpad API now allows copies of unpublished packages, I could remove the check from ppa-copy-packages and get faster turnaround times for my package publishing!

**Comment by benjaoming** [comment]
I tried it many many times, the binaries in the source dist were already published days back.

I'm afraid that I couldn't understand why this was failing, that's why I resorted to asking if you had any issues currently.

**Comment by mgedmin** [comment]
Hm, this is interesting.

I wonder if there's some package name confusion in the code?  I haven't tried ppa-copy-packages on packages where the source package name differs from binary package names (or even on packages where one source package builds multiple binary packages).

The code that prints that warning is here:
https://github.com/mgedmin/ppa-copy-packages/blob/d4de9fca8ccb1147dee4a31ad628977d4553fe28/ppa_copy_packages.py#L350-L360

and I find it suspicious that it checks `builds[0]` but ignores all other builds.

Actually wait, that's just for printing a warning.  The check itself is
https://github.com/mgedmin/ppa-copy-packages/blob/d4de9fca8ccb1147dee4a31ad628977d4553fe28/ppa_copy_packages.py#L344-L346

and this is where the bug lies!

https://github.com/mgedmin/ppa-copy-packages/blob/d4de9fca8ccb1147dee4a31ad628977d4553fe28/ppa_copy_packages.py#L264-L274

The code is assuming there's a binary package with the same name as the source package!

**Comment by mgedmin** [comment]
Here's what the script needs to do: given a source package name + version, it needs to figure out if all the binary packages built from it have already been published.

Here's the API documentation: https://launchpad.net/+apidoc/devel.html

It's like a puzzle!  But my brain is no longer working today.

https://launchpad.net/+apidoc/devel.html#binary_package_publishing_history has no 'source_package_name' field, which would make this easy, or at least easier.

The old solution assumed we could batch binary and source publication records by package name and version.

I think it may be a better idea to match by build_link?  [Builds](https://launchpad.net/+apidoc/devel.html#build) have a source_package_name field.  This would solve half the problem: how to check if _some_ packages built from a particular source package version have been published yet.  I'm not sure how to check whether _all_ packages built from the same source package have been built.

Ideally I'd prefer to fetch the build for the right source package in advance, so I could compare the build_link instead of fetching O(N) builds for each published binary package and looking at the package names inside.  Launchpad API is very slow: 11 API requests take about 5 seconds.
If you run

    ppa-copy-packages --dry-run --owner learningequality --name ka-lite-proposed -p ka-lite-source -s trusty -t focal -vv

you can look at the timestamps (+deltas) and [number of requests/+delta], and you'll see that each request takes between 200ms and 1000ms.
